### PR TITLE
Dict-like space

### PR DESCRIPTION
**Pull request by nfcampos**
this PR adds a dict-like Space class (closes issue #140 )

changes: 
- added `DictSpace` class
- added `make_space` function for turning dimensions passed to `*_minimize` into a Space
- added tests for `DictSpace`
- added tests for `make_space`
- updated hyperparameter optimization example to use a dict space (it's much better for passing parameters to a scikit estimator)

questions:
- do you guys want the `dimensions` property of a dict space to be a dictionary or a list? for now I kept it as a list, let me know if i should change that (whatever is decided for this should will also decide whether `bounds` are a dict or a list)

### CODE REVIEW

File: skopt/space.py [hunk 1]
```diff
@@ -486,3 +486,39 @@ def transformed_bounds(self):
...
+    def inverse_transform(self, Xt):
+        return [
+            dict(zip(self.keys, sample))
+            for sample in super(DictSpace, self).inverse_transform(Xt)
+        ]
+
+
+def make_space(dimensions):
+    if isinstance(dimensions, Space):
+        return dimensions
```
Comment (nfcampos) [inline skopt/space.py]: with this change it is now possible to pass to *_minimize a `Space` instance directly, which was not possible before, let me know if you don't like that for some reason and I can remove this branch from the function

**Comment by betatim** [comment]
Travis is unhappy 😢. Some quick thoughts/questions:

* why do we need `DictSpace`? Can't `make_space` take care of the different ways of specifying a space and create `Space` instances for all of them?
* if we get into modifying how to specify spaces we should consider how to specify (in the future, not this PR) a conditional space. Say one dimension is: "no preprocessing" and "PCA" then if you choose "PCA" there will be extra dimensions that don't exist if you choose "no preprocessing".
* I like that we can now specify names for dimensions, I think the names should be stored in the `Dimension` object

I will take a closer look this evening.

**Comment by nfcampos** [comment]
@betatim 
- re: Travis being unhappy, I didn't realise some tests define `dimensions` as a numpy array, I've added explicit support for that to `make_space`
- re: the point of `DictSpace` is that its samples are `dict`s instead of `list`s, which in lots of cases is much better for use inside the objective function
- re: conditional spaces I wrote something for that some weeks ago but never opened a PR, it's now here #199
  - this only supports conditional spaces where the dimensions are always the same but some combinations of values are invalid

**Comment by nfcampos** [comment]
do you guys have any interest in something like this, ie. should I resolve the conflicts with the latest master?

**Comment by pavelkomarov** [comment]
Should we be deleting these old pull requests that never got merged?

**Comment by nfcampos** [comment]
I won’t be updating this so I’m closing it myself.

**Pull request closed without merging**